File: win/Sources/CameraService.cs
using System;
using System.Linq;
using System.Threading.Tasks;
using Windows.Devices.Enumeration;
using Windows.Graphics.Imaging;
using Windows.Media.Capture;
using Windows.Media.Capture.Frames;
using Windows.Media.MediaProperties;
using Windows.Storage.Streams;

namespace GlassFace;

/// <summary>A single BGRA8 camera frame (tightly packed, stride = Width * 4).</summary>
public sealed class CameraFrame
{
    public byte[] Bgra { get; }
    public int Width { get; }
    public int Height { get; }

    public CameraFrame(byte[] bgra, int width, int height)
    {
        Bgra = bgra;
        Width = width;
        Height = height;
    }
}

/// <summary>
/// Receives raw camera frames. Implemented by whatever wants to render or analyze them
/// (here, the overlay) — <see cref="CameraService"/> depends on this abstraction, not the
/// concrete type. Mirrors the macOS <c>FrameConsumer</c> protocol.
/// </summary>
public interface IFrameConsumer
{
    /// <summary>Called on a capture thread (off the UI thread).</summary>
    void Consume(CameraFrame frame);
}

/// <summary>
/// Owns the WinRT <see cref="MediaCapture"/> graph: device discovery, the frame reader,
/// and starting/stopping capture. Its single responsibility is supplying camera frames as
/// BGRA8 to its <see cref="FrameConsumer"/>. Knows nothing about windows or layers.
/// </summary>
public sealed class CameraService
{
    public IFrameConsumer? FrameConsumer { get; set; }

    private MediaCapture? _capture;
    private MediaFrameReader? _reader;

    private readonly object _gate = new();
    private byte[]? _buffer;                       // reused across frames to avoid churn
    private Windows.Storage.Streams.Buffer? _winrtBuffer;
    private bool _running;

    /// <summary>
    /// Builds the capture graph. Throws <see cref="UnauthorizedAccessException"/> if the user
    /// has denied camera access, or <see cref="NotSupportedException"/> if no camera is usable.
    /// </summary>
    public async Task ConfigureAsync()
    {
        var device = await FrontCameraAsync();    // throws NotSupportedException if none

        var settings = new MediaCaptureInitializationSettings
        {
            VideoDeviceId = device.Id,
            StreamingCaptureMode = StreamingCaptureMode.Video,
            MemoryPreference = MediaCaptureMemoryPreference.Cpu,   // CPU frames -> easy to blit
            SharingMode = MediaCaptureSharingMode.ExclusiveControl
        };

        _capture = new MediaCapture();
        await _capture.InitializeAsync(settings);  // triggers the OS camera-consent check

        var source = _capture.FrameSources.Values.FirstOrDefault(s =>
                         s.Info.SourceKind == MediaFrameSourceKind.Color)
                     ?? throw new NotSupportedException("No color camera frame source available.");

        _reader = await _capture.CreateFrameReaderAsync(source, MediaEncodingSubtypes.Bgra8);
        _reader.AcquisitionMode = MediaFrameReaderAcquisitionMode.Realtime;  // drop late frames
        _reader.FrameArrived += OnFrameArrived;
    }

    public Task StartAsync() => SetRunningAsync(true);

    /// <summary>Starts/stops the reader. Stopping turns the camera (and its privacy light) off.</summary>
    public async Task SetRunningAsync(bool running)
    {
        if (_reader is null) return;
        if (running && !_running)
        {
            await _reader.StartAsync();
            _running = true;
        }
        else if (!running && _running)
        {
            await _reader.StopAsync();
            _running = false;
        }
    }

    private void OnFrameArrived(MediaFrameReader sender, MediaFrameArrivedEventArgs args)
    {
        using var frame = sender.TryAcquireLatestFrame();
        var bitmap = frame?.VideoMediaFrame?.SoftwareBitmap;
        if (bitmap is null) return;

        // Normalize to opaque BGRA8 (matches the WriteableBitmap's Bgr32 layout).
        SoftwareBitmap? converted = null;
        if (bitmap.BitmapPixelFormat != BitmapPixelFormat.Bgra8 ||
            bitmap.BitmapAlphaMode != BitmapAlphaMode.Ignore)
        {
            converted = SoftwareBitmap.Convert(bitmap, BitmapPixelFormat.Bgra8, BitmapAlphaMode.Ignore);
            bitmap = converted;
        }

        try
        {
            int w = bitmap.PixelWidth, h = bitmap.PixelHeight, size = w * h * 4;
            lock (_gate)
            {
                if (_buffer is null || _buffer.Length != size)
                {
                    _buffer = new byte[size];
                    _winrtBuffer = new Windows.Storage.Streams.Buffer((uint)size);
                }

                bitmap.CopyToBuffer(_winrtBuffer);
                using var reader = DataReader.FromBuffer(_winrtBuffer);
                reader.ReadBytes(_buffer);

                // Consume synchronously (see OverlayCoordinator) so _buffer is safe to reuse.
                FrameConsumer?.Consume(new CameraFrame(_buffer, w, h));
            }
        }
        finally
        {
            converted?.Dispose();
        }
    }

    /// <summary>Prefers the front-facing (selfie) camera; falls back to any available camera.</summary>
    private static async Task<DeviceInformation> FrontCameraAsync()
    {
        var devices = await DeviceInformation.FindAllAsync(DeviceClass.VideoCapture);
        if (devices.Count == 0)
            throw new NotSupportedException("No camera device found.");

        return devices.FirstOrDefault(d => d.EnclosureLocation?.Panel == Panel.Front)
               ?? devices[0];
    }
}

File: win/Sources/CameraService.cpp
#include "pch.h"
#include "CameraService.h"

using namespace winrt;
using namespace winrt::Windows::Foundation;
using namespace winrt::Windows::Foundation::Collections;
using namespace winrt::Windows::Devices::Enumeration;
using namespace winrt::Windows::Graphics::Imaging;
using namespace winrt::Windows::Media::Capture;
using namespace winrt::Windows::Media::Capture::Frames;
using namespace winrt::Windows::Media::MediaProperties;

namespace glassface
{
    IAsyncAction CameraService::ConfigureAsync()
    {
        // Prefer the front-facing (selfie) camera; fall back to any available camera.
        auto devices = co_await DeviceInformation::FindAllAsync(DeviceClass::VideoCapture);
        if (devices.Size() == 0)
            throw hresult_error(E_FAIL, L"No camera device found.");

        DeviceInformation chosen{ nullptr };
        for (auto const& device : devices)
        {
            auto enclosure = device.EnclosureLocation();
            if (enclosure && enclosure.Panel() == Panel::Front) { chosen = device; break; }
        }
        if (!chosen) chosen = devices.GetAt(0);

        auto settings = MediaCaptureInitializationSettings();
        settings.VideoDeviceId(chosen.Id());
        settings.StreamingCaptureMode(StreamingCaptureMode::Video);
        settings.MemoryPreference(MediaCaptureMemoryPreference::Cpu);   // CPU frames -> easy to display
        settings.SharingMode(MediaCaptureSharingMode::ExclusiveControl);

        m_capture = MediaCapture();
        co_await m_capture.InitializeAsync(settings);   // throws E_ACCESSDENIED if user declined

        MediaFrameSource source{ nullptr };
        for (auto const& kv : m_capture.FrameSources())
        {
            if (kv.Value().Info().SourceKind() == MediaFrameSourceKind::Color)
            {
                source = kv.Value();
                break;
            }
        }
        if (!source)
            throw hresult_error(E_FAIL, L"No color camera frame source available.");

        m_reader = co_await m_capture.CreateFrameReaderAsync(source, MediaEncodingSubtypes::Bgra8());
        m_reader.AcquisitionMode(MediaFrameReaderAcquisitionMode::Realtime);   // drop late frames
        m_frameToken = m_reader.FrameArrived({ this, &CameraService::OnFrameArrived });
    }

    IAsyncAction CameraService::StartAsync()
    {
        co_await SetRunningAsync(true);
    }

    IAsyncAction CameraService::SetRunningAsync(bool running)
    {
        if (!m_reader) co_return;
        if (running && !m_running)
        {
            co_await m_reader.StartAsync();
            m_running = true;
        }
        else if (!running && m_running)
        {
            co_await m_reader.StopAsync();
            m_running = false;
        }
    }

    void CameraService::OnFrameArrived(MediaFrameReader const& sender, MediaFrameArrivedEventArgs const&)
    {
        auto frame = sender.TryAcquireLatestFrame();
        if (!frame) return;
        auto videoFrame = frame.VideoMediaFrame();
        if (!videoFrame) return;
        auto bitmap = videoFrame.SoftwareBitmap();
        if (!bitmap) return;

        // SoftwareBitmapSource wants BGRA8 with premultiplied alpha.
        if (bitmap.BitmapPixelFormat() != BitmapPixelFormat::Bgra8 ||
            bitmap.BitmapAlphaMode() != BitmapAlphaMode::Premultiplied)
        {
            bitmap = SoftwareBitmap::Convert(bitmap, BitmapPixelFormat::Bgra8, BitmapAlphaMode::Premultiplied);
        }

        if (frameConsumer)
            frameConsumer(bitmap);
    }
}
